### PR TITLE
[Merged by Bors] - feat: basic definitions for Algebra.Homology.ShortComplex.LeftHomology

### PR DESCRIPTION
**Pull request by joelriou**
This PR develops notion of `LeftHomologyData` for a short complex involving two maps `f` and `g`, where homology is understood as a quotient (cokernel) of the kernel of the second map `g`. It also introduces the structure `LeftHomologyMapData` which will allow computations of induced maps on (left) homology.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean [hunk 1]
```diff
@@ -61,6 +60,206 @@ structure LeftHomologyData where
...
+attribute [reassoc (attr := simp)] wi wπ
+
+variable {S}
+variable (h : S.LeftHomologyData) {A : C}
+
+instance : Mono h.i := ⟨fun _ _ => Fork.IsLimit.hom_ext h.hi⟩
+
+instance : Epi h.π := ⟨fun _ _ => Cofork.IsColimit.hom_ext h.hπ⟩
+
+/-- any morphism `k : A ⟶ S.X₂` that is a cycle (i.e. `k ≫ S.g = 0`) lifts
```
Comment (jcommelin) [inline Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean:88]: I notice that some of your docstrings start with a lowercase letter, others with uppercase. Is there a pattern?
Anyway, I think I prefer that all docstrings start with an uppercase letter.

Comment (joelriou) [reply Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean]: Ok, I have changed to Uppercase for the definitions, but kept lowercase for docstrings of fields of structures.

File: Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean [hunk 2]
```diff
@@ -61,6 +60,206 @@ structure LeftHomologyData where
...
+noncomputable def leftHomologyData [S.HasLeftHomology] :
+  S.LeftHomologyData := HasLeftHomology.condition.some
+
+variable {S}
+
+namespace HasLeftHomology
+
+lemma mk' (h : S.LeftHomologyData) : HasLeftHomology S := ⟨Nonempty.intro h⟩
+
+instance of_kerKernel_of_hasCokernel [HasKernel S.g] [HasCokernel (kernel.lift S.g S.f S.zero)] :
```
Comment (semorrison) [inline Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean:218]:
```suggestion
instance of_hasKernel_of_hasCokernel [HasKernel S.g] [HasCokernel (kernel.lift S.g S.f S.zero)] :
```

File: Mathlib/CategoryTheory/Limits/Shapes/Kernels.lean [hunk 3]
```diff
@@ -193,6 +193,18 @@ def isKernelOfComp {W : C} (g : Y ⟶ W) (h : X ⟶ W) {c : KernelFork h} (i : I
...
 
+/-- `X` identifies to the kernel of a zero map `X ⟶ Y`. -/
+def KernelFork.IsLimit.ofId {X Y : C} (f : X ⟶ Y) (hf : f = 0) :
+    IsLimit (KernelFork.ofι (𝟙 X) (show 𝟙 X ≫ f = 0 by rw [hf, comp_zero])) :=
+  KernelFork.IsLimit.ofι _ _ (fun x _ => x) (fun _ _ => Category.comp_id _)
+    (fun _ _ _ hb => by simp only [← hb, Category.comp_id])
+
+/-- Any zero object identifies to the kernel of a given monomorphisms. -/
+def KernelFork.IsLimit.ofIsZeroOfMono {X Y : C} {f : X ⟶ Y} (c : KernelFork f)
+    (hf : Mono f) (h : IsZero c.pt) : IsLimit c :=
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Limits/Shapes/Kernels.lean:204]: Swap order of arguments or rename to `ofMonoOfIsZero`? And below.

Comment (joelriou) [reply Mathlib/CategoryTheory/Limits/Shapes/Kernels.lean]: Thanks for the suggestions! I have just changed the names.

File: Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean [hunk 4]
```diff
@@ -61,6 +60,206 @@ structure LeftHomologyData where
...
+  hπ := CokernelCofork.IsColimit.ofId _ hf
+
+@[simp] lemma ofZeros_f' (hf : S.f = 0) (hg : S.g = 0) :
+    (ofZeros S hf hg).f' = 0 := by
+  rw [← cancel_mono ((ofZeros S hf hg).i), zero_comp, f'_i, hf]
+
+end LeftHomologyData
+
+/-- A short complex `S` has left homology when there exists a `S.LeftHomologyData` -/
+class HasLeftHomology : Prop where
```
Comment (jcommelin) [inline Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean:205]: This class is mentioned as `TODO` in the module docstring at the top of the file. Should that docstring now be updated?

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**